src/util/valuetransformer.cpp

```diff
@@ -1,35 +1,130 @@
+    }
+
+  private:
+    double m_addend;
+};
+
+class TransformInvert : public TransformNode {
+  public:
+    double transform(double argument) const override {
+        return -argument;
+    }
+
+    double transformInverse(double argument) const override {
+        return -argument;
+    }
+};
+
+class TransformNot : public TransformNode {
+  public:
+    double transform(double argument) const override {
+        return !static_cast<bool>(argument);
+    }
 
-void ValueTransformer::addTransformer(TransformNode* pTransformer) {
-    m_transformers.append(pTransformer);
+    double transformInverse(double argument) const override {
+        return !static_cast<bool>(argument);
+    }
+};
+
+class TransformIsEqual : public TransformNode {
+  public:
+    TransformIsEqual(double compareValue)
+            : m_compareValue(compareValue) {
+    }
+
+    double transform(double argument) const override {
+        return argument == m_compareValue;
+    }
+
+    double transformInverse(double argument) const override {
+        if (argument > 0.0) {
+            return m_compareValue;
+        }
+        return 0.0;
+    }
+
+  private:
+    double m_compareValue;
+};
+
+} // namespace
+
+void ValueTransformer::addTransformer(std::unique_ptr<TransformNode> pTransformer) {
+    m_transformers.push_back(std::move(pTransformer));
 }
 
 double ValueTransformer::transform(double argument) const {
-    foreach (const TransformNode* pNode, m_transformers) {
-        argument = pNode->transform(argument);
-    }
-    return argument;
+    return std::accumulate(m_transformers.cbegin(),
+            m_transformers.cend(),
+            argument,
+            [&](double argument, const auto& pNode) {
+                return pNode->transform(argument);
+            });
 }
 
 double ValueTransformer::transformInverse(double argument) const {
-    for (int i = m_transformers.size() - 1; i >= 0; --i) {
-        const TransformNode* pNode = m_transformers[i];
-        argument = pNode->transformInverse(argument);
-    }
-    return argument;
+    return std::accumulate(m_transformers.crbegin(),
+            m_transformers.crend(),
+            argument,
+            [&](double argument, const auto& pNode) {
+                return pNode->transformInverse(argument);
+            });
 }
 
 // static
-ValueTransformer* ValueTransformer::parseFromXml(const QDomElement& transformElement,
-        const SkinContext& context) {
+std::unique_ptr<ValueTransformer> ValueTransformer::parseFromXml(
+        const QDomElement& transformElement, const SkinContext& context) {
     if (transformElement.isNull() || !transformElement.hasChildNodes()) {
         return nullptr;
     }
 
-    ValueTransformer* pTransformer = new ValueTransformer();
+    // constructor is private, so we can't use std::make_unique
+    auto pTransformer = std::unique_ptr<ValueTransformer>(new ValueTransformer());
+
+    auto maybeAddFromElement = [&]<typename T>(auto element) {
```

Comment:
Is there a reason this cannot be a normal template function? This definition and the calls below look quite uncommon. 
How about: 
`pTransformer->maybeAddFromElement<typename T>(QDomElement& element);`

Reply:
done. please review.